ShaiRandom/Generators/IEnhancedRandomExtensions.cs

```diff
@@ -253,63 +265,302 @@ public static float MaxFloatOf(this IEnhancedRandom rng, float innerBound, float
+        /// </remarks>
+        /// <exception cref="ArgumentException">An empty span was given.</exception>
+        /// <param name="rng" />
+        /// <typeparam name="T">The type of items in the span.</typeparam>
+        /// <param name="items">Must be non-empty.</param>
+        /// <param name="selector">Function that should return true _only_ for a valid item to select.</param>
+        /// <returns>A randomly-chosen item from the span for which <see cref="selector"/> returns true.</returns>
+        public static T RandomElement<T>(this IEnhancedRandom rng, ReadOnlySpan<T> items, Func<T, bool> selector)
+        {
+            T item = rng.RandomElement(items);
+            while (!selector(item))
+                item = rng.RandomElement(items);
+
+            return item;
+        }
+
+        /// <summary>
+        /// Continuously selects random items from the given non-empty span, until either one is found for which
+        /// <paramref name="selector"/> returns true, or the maximum number of attempts is reached.  An exception will
+        /// be thrown if the function is unable to generate a valid item within the number of attempts specified.
+        /// </summary>
+        /// <exception cref="ArgumentException">An empty span was given.</exception>
+        /// <exception cref="MaxAttemptsReachedException">The function could not find a valid element within the maximum number of tries.</exception>
+        /// <param name="rng" />
+        /// <typeparam name="T">The type of items in the span.</typeparam>
+        /// <param name="items">Must be non-empty.</param>
+        /// <param name="selector">Function that should return true _only_ for a valid item to select.</param>
+        /// <param name="maxTries">Maximum number of times to try generating a valid value before giving up and throwing an exception.</param>
+        /// <returns>A randomly-chosen item from the span for which <see cref="selector"/> returns true.</returns>
+        public static T RandomElement<T>(this IEnhancedRandom rng, ReadOnlySpan<T> items, Func<T, bool> selector, int maxTries)
+        {
+            if (maxTries <= 0)
+                throw new ArgumentOutOfRangeException(nameof(maxTries),
+                    "Value must >= 0; for infinite retries, use the overload without a maxTries parameter.");
+
+            int curTries = 0;
+            while (curTries < maxTries)
+            {
+                T item = rng.RandomElement(items);
+                if (selector(item))
+                    return item;
+
+                curTries++;
+            }
+
+            throw new MaxAttemptsReachedException();
+        }
 
         /// <summary>
         /// Gets a randomly-chosen item from the given non-null, non-empty IReadOnlyList.
         /// </summary>
+        /// <remarks>
+        /// Note that this function can take arrays as well as the <paramref name="items"/> value, since arrays implement
+        /// IReadOnlyList.
```

Comment:
I like having this noted, and I hope users get a chance to see it before they get confused about getting random items from arrays.